### PR TITLE
implemented 'torch.distributions.constraints.symmetric' checking if the tensor is symmetric at last 2 dimension.

### PR DESCRIPTION
**Pull request by nonconvexopt**
Implemented submodule for #68050
Opened cleaned, final version of PR for #68240

Explanation:
I am trying to contribute to PyTorch by implementing distributions for symmetric matrices like Wishart distribution and Inverse Wishart distribution. Although there is a LKJ distribution for the Cholesky decomposition of correlation matrices, it only represents equivalence to restricted form of Wishart distribution. [https://arxiv.org/abs/1809.04746](https://arxiv.org/abs/1809.04746) Thus, I started implementing Wishart distribution and Inverse Wishart distribution seperately.

I added a short code about the 'torch.distributions.constraints.symmetric', which was not included in 'torch.distributions.constraints' previously. i.e., 'torch.distributions.constraints' contains module like 'positive_definite' constraints, but it just assumes symmetricity of the input matrix. [Link](https://github.com/pytorch/pytorch/blob/1adeeabdc0c8832420c091c5c668843768530d7f/torch/distributions/constraints.py#L466) So, I think it will be better if we have constraint checking symmetricity of the tensors in PyTorch.

We may further utilize it like
`constraints.stack([constraints.symmetric, constraints.positive_definite])`
for the constraint of the covariance matrix in Multivariate Normal distribution, for example, to check if the random matrix is a symmetric positive definite matrix.

cc @fritzo @neerajprad @alicanb @nikitaved

### CODE REVIEW

**Comment by neerajprad** [comment]
Looks great. Could you add as small test to [test_constraints](https://github.com/pytorch/pytorch/blob/master/test/distributions/test_constraints.py) which tests this on a few matrices and verifies that this can be stacked with `positive_definite` so that matrices like `[[2., 0.], [2., 2.]]` will fail this constraint but pass the positive definite constraint?

**Comment by nonconvexopt** [comment]
> Looks great. Could you add as small test to [test_constraints](https://github.com/pytorch/pytorch/blob/master/test/distributions/test_constraints.py) which tests this on a few matrices and verifies that this can be stacked with `positive_definite` so that matrices like `[[2., 0.], [2., 2.]]` will fail this constraint but pass the positive definite constraint?

Sure, it is my pleasure. Thank you for giving me detailed advice. I will test it out soon!

**Comment by nonconvexopt** [comment]
> Looks great. Could you add as small test to [test_constraints](https://github.com/pytorch/pytorch/blob/master/test/distributions/test_constraints.py) which tests this on a few matrices and verifies that this can be stacked with `positive_definite` so that matrices like `[[2., 0.], [2., 2.]]` will fail this constraint but pass the positive definite constraint?

@neerajprad How about this design? [inheritable_torch.distributions.constraints_for_matrices](https://github.com/nonconvexopt/pytorch/blob/b371797234fa2bb59e9f455b8530ba8134e557c7/torch/distributions/constraints.py#L463-L491) It is different from previous implementation styles in 'torch.distributions.constraints' and it requires more computation. But more logical way.

**Comment by nonconvexopt** [comment]
This is the best way I think. I followed the design styles from 'test/distributions/test_distributions.py' which used 'unittest' package rather than 'pytest' package.
Also, I found an error in the implementation of 'torch.distributions.constraints.positive_definite' which I will make another PR.

**Comment by neerajprad** [comment]
Looks good to me! Thanks for contributing, I'll review your PR for the Wishart distribution after this.

File: test/distributions/test_constraints.py [hunk 1]
```diff
@@ -41,6 +48,15 @@ def build_constraint(constraint_fn, args, is_cuda=False):
     t = torch.cuda.DoubleTensor if is_cuda else torch.DoubleTensor
     return constraint_fn(*(t(x) if isinstance(x, list) else x for x in args))
 
+@pytest.mark.parametrize('constraint_fn, result, value', [(e[0], e[1], e[2]) for e in EXAMPLES])
```
Comment (neerajprad) [inline test/distributions/test_constraints.py:51]: nit: why not just `@pytest.mark.parametrize('constraint_fn, result, value', EXAMPLES)`?

**Review by neerajprad — APPROVED** [review]

File: test/distributions/test_constraints.py (hunk 1 above)
Comment (nonconvexopt) [reply test/distributions/test_constraints.py:51]: You are right. Thanks for checking!

**Comment by nonconvexopt** [comment]
CI errors are caused due to the tests about constraints.positive_definite:
`(constraints.positive_definite, True, [[2., 0], [2., 2]]),`
This is the friendly reminder, explanation from Lezcano: [Debug positive definite constraints](https://github.com/pytorch/pytorch/pull/68720#pullrequestreview-813311457)
We can remove the test to pass the CI test.

**Comment by neerajprad** [comment]
> This is the friendly reminder, explanation from Lezcano: Debug positive definite constraints
We can remove the test to pass the CI checks.

I didn't see any failures related to this. Could you point me to the test errors, are the failures deterministic? We should definitely skip the test (or add `pytest.xfail`) if this is expected to fail with CUDA.

**Comment by ngimel** [comment]
Breaks cuda 10.2 build, reverting, see e.g. https://github.com/pytorch/pytorch/runs/4318116960?check_suite_focus=true

**Pull request closed without merging**